bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/watson/ElementTree.java

```diff
@@ -79,65 +80,70 @@
  * 	- "It's ElementTree my dear Watson, ElementTree."
  */
 public class ElementTree {
-	protected DeltaDataTree tree;
-	protected IElementTreeData userData;
+	protected final DeltaDataTree tree;
+	protected volatile IElementTreeData userData;
 
-	private static class ChildIDsCache {
+	private static final class ChildIDsCache {
 		ChildIDsCache(IPath path, IPath[] childPaths) {
 			this.path = path;
 			this.childPaths = childPaths;
 		}
 
-		IPath path;
-		IPath[] childPaths;
+		final IPath path;
+		final IPath[] childPaths;
 	}
 
+	/** synchronized access **/
 	private volatile ChildIDsCache childIDsCache = null;
 
+	/** synchronized access **/
 	private volatile DataTreeLookup lookupCache = null;
 
+	/** synchronized access **/
 	private volatile DataTreeLookup lookupCacheIgnoreCase = null;
 
-	private static int treeCounter = 0;
-	private int treeStamp;
+	private final static AtomicInteger treeCounter = new AtomicInteger();
+	private final int treeStamp;
 
 	/**
 	 * Creates a new empty element tree.
 	 */
 	public ElementTree() {
-		initialize(new DeltaDataTree());
+		this(new DeltaDataTree());
 	}
 
 	/**
 	 * Creates an element tree given its internal node representation.
 	 */
 	protected ElementTree(DataTreeNode rootNode) {
-		initialize(rootNode);
+		/* create the implicit root node */
+		this(new DeltaDataTree(new DataTreeNode(null, null, new AbstractDataTreeNode[] { rootNode })));
 	}
 
 	/**
 	 * Creates a new element tree with the given data tree as its representation.
 	 */
-	protected ElementTree(DeltaDataTree tree) {
-		initialize(tree);
+	protected ElementTree(DeltaDataTree newTree) {
+		// Keep this element tree as the data of the root node.
+		// Useful for canonical results for ElementTree.getParent().
+		// see getParent().
+		treeStamp = treeCounter.incrementAndGet();
+		newTree.setData(newTree.rootKey(), this);
+		this.tree = newTree;
 	}
 
 	/**
 	 * Creates a new empty delta element tree having the
 	 * given tree as its parent.
 	 */
 	protected ElementTree(ElementTree parent) {
-		if (!parent.isImmutable()) {
-			parent.immutable();
```

Comment:
why is this now removed?

Reply:
that moved to the caller  newEmptyDelta():624 to get the constructor at the first place.